### PR TITLE
fix `teleport db configure create`

### LINKED ISSUE #19560: teleport db configure create flag checking is inconsistent with valid database config
Expected behavior:
I should be able to create a config for a static database that doesn't require a URI e.g. AWS Keyspaces (Cassandra):
```sh
$ teleport db configure create \ 
      --name="foo" \
      --proxy="mars.local.gd:443" \
      --protocol=cassandra \
      --aws-region="us-west-1" \
      --token="abc123"
```
Current behavior:
prints
`ERROR: --uri is required when configuring static database`

Furthermore, `--aws-accound-id` (for static database config) is missing from the config flags. This setting is required for AWS Keyspaces.

I think we should have very few flag checks in this command and instead we should generate the yaml config, parse it, and then try to build a database from it. This way we can keep this command's flag validation in sync with the database config spec validation code.

Bug details:
- Teleport version: dev v12  (4237ad3cc72176e343684f3b807c6b2cc11e2968)

@smallinsky

### PR DESCRIPTION
**Pull request by GavinFrazar**
Fixes #19560

## What
This change synchronizes `teleport db configure create` parameter validation with the checks we do when actually running the generated configuration. This way we don't generate configs that are invalid and we don't throws errors in `teleport db configure create` for configurations that are actually valid.

## How
When we generate a db agent config, we immediately parse the config and apply it to a default configuration (without actually running the configuration), and then if that doesn't throw any errors we dump the config like we did before. This way we share all the same validation logic as when actually running the configuration.

While doing this it became necessary to add several missing flags to the command and fix some issues with the config generation template as well.

I updated the test cases to cover all config flags at least once, and a few cases for configurations that are valid but were rejected without this change (such as cassandra/dynamodb with URI derived by AWS region instead of specifying it).

### CODE REVIEW

**Comment by GavinFrazar** [comment]
@smallinsky it might be nice to do the same for other Teleport services. I haven't investigated if that leads to any issues though. Database agent config is relatively simple compared to other services. Do you think we should open a ticket for other services to do the same (i.e. do a "dry-run" with config from `teleport configure`)?

File: lib/config/database.go
Comment (greedy52) [inline lib/config/database.go:596]: nice 👍

File: tool/teleport/common/teleport.go
Comment (GavinFrazar) [reply tool/teleport/common/teleport.go:244]: In my opinion, we should encourage the `teleport db configure create... teleport start` flow in our docs, but still have the flags on `teleport start`. I think every configuration value should have a corresponding imperative flag in our CLI for scripting purposes. But that can be a discussion for the CLI guidelines RFD when we get to that, so I can remove it from this PR for now.

File: tool/teleport/common/teleport.go
Comment (GavinFrazar) [reply tool/teleport/common/teleport.go:289]: I'd prefer keeping them for consistency and to prevent a scenario where a configuration value can't be specified but it's required to pass the config verification checks.

Re: `(Optional for XXX)` vs `(Required for YYY)` - I think those cases will naturally become obvious when running the command. It may be that a flag is sometimes required, but depends on the settings for other flags. For instance, `--uri` is technically optional for DynamoDB, but only if `--aws-region` is specified. So determining optional vs required upfront in the help string isn't really possible to do, at least not accurately for all cases.

I definitely missed adding a lot of values from `lib/config/fileconf.Database` though, and I'd prefer we have them all...

Comment (greedy52) [reply tool/teleport/common/teleport.go:289]: > So determining optional vs required upfront in the help string isn't really possible to do, at least not accurately for all cases.

Agreed. Though when I read through `teleport db configure create -h`, it feels to me that one of those IDs is required. Ideally, we want to save the time that spend on grabbing these ids. Could be me just me though.

File: api/types/database.go
Comment (GavinFrazar) [reply api/types/database.go:492]: Formatting of user messages depends on how the error is logged and what format the logger is using.

I went into my cluster's sql backend and modified a dynamic database manually to produce this error, because I'm actually not sure how it can be created in the backend otherwise.

We add an additional user message "failed to fetch resource: %q" here: https://github.com/gravitational/teleport/blob/517f93e1b42e76b400f796cef58ba9bd23858578/lib/cache/cache.go#L1287-L1290

and we log the error using `WithError` here: https://github.com/gravitational/teleport/blob/517f93e1b42e76b400f796cef58ba9bd23858578/lib/cache/cache.go#L849-L851

JSON logging example:
```
{
  "caller": "cache/cache.go:850",
  "component": "auth:1:cache",
  "error": "failed to fetch resource: \"db\"\n\tdatabase \"roach\"\n\t\tinvalid origin value \"bad-config\", must be one of [defaults config-file dynamic cloud kubernetes]",
  "level": "warning",
  "message": "Re-init the cache on error",
  "timestamp": "2023-01-31T11:48:08-08:00"
}
```

text example:
```
2023-01-31T11:52:46-08:00 WARN [AUTH:1:CA] Re-init the cache on error error:[
ERROR REPORT:
Original Error: *trace.BadParameterError invalid origin value &#34;bad-config&#34;, must be one of [defaults config-file dynamic cloud kubernetes]
Stack Trace:
	github.com/gravitational/teleport/api@v0.0.0/types/resource.go:413 github.com/gravitational/teleport/api/types.(*Metadata).CheckAndSetDefaults
	github.com/gravitational/teleport/api@v0.0.0/types/database.go:491 github.com/gravitational/teleport/api/types.(*DatabaseV3).CheckAndSetDefaults
	github.com/gravitational/teleport/lib/services/database.go:123 github.com/gravitational/teleport/lib/services.UnmarshalDatabase
	github.com/gravitational/teleport/lib/services/local/databases.go:48 github.com/gravitational/teleport/lib/services/local.(*DatabaseService).GetDatabases
	github.com/gravitational/teleport/lib/cache/collections.go:1465 github.com/gravitational/teleport/lib/cache.(*database).fetch
	github.com/gravitational/teleport/lib/cache/cache.go:1287 github.com/gravitational/teleport/lib/cache.(*Cache).fetch.func2
	golang.org/x/sync@v0.1.0/errgroup/errgroup.go:75 golang.org/x/sync/errgroup.(*Group).Go.func1
	runtime/asm_arm64.s:1172 runtime.goexit
User Message: failed to fetch resource: &#34;db&#34;
	database &#34;roach&#34;
		invalid origin value &#34;bad-config&#34;, must be one of [defaults config-file dynamic cloud kubernetes]] cache/cache.go:850
```

I think it's appropriate to add a user message that includes the name of the database so it's clear which one is the problem. We just need to fix our log formatting to handle user messages better (among other problems). There's an RFD for that here: https://github.com/gravitational/teleport/blob/428eb9eff08478e1758da2ce90aac99df079b001/rfd/0092-error-message-guidelines.md

File: tool/teleport/common/teleport.go
Comment (GavinFrazar) [reply tool/teleport/common/teleport.go:289]: Maybe we should indicate where the flag applies to static db config only? After all with these AWS databases, we prefer users to setup discovery right?

Comment (greedy52) [reply tool/teleport/common/teleport.go:289]: Ya, discovery should be preferred. Though the "ids" are optional even for static dbs. Only need the right AWS `--uri`. The "ids" are usually redundant. 

Some Azure static dbs do require certain fields though, sigh.

File: api/types/database.go
Comment (greedy52) [reply api/types/database.go:492]: Definitely better with the database name included. Do you know if we have plan to change the UserMessage format?

I was curious about this only because I am not a big fan of having new lines in the message. I would probably prefer embed the name in the same message or wrap the name in the caller like:
https://github.com/gravitational/teleport/blob/26a623b5be96a8766541fabf01a232b3e57f1a82/lib/config/configuration.go#L1373-L1374

Comment (GavinFrazar) [reply api/types/database.go:492]: > Do you know if we have plan to change the UserMessage format?
I think that will happen after the RFD is approved. I have a comment here: https://github.com/gravitational/teleport/pull/17305#discussion_r1073094451

For text formatting, I would like to get rid of the HTML escaping. There's a setting for that, I'm just not sure about the full implications of disabling that escaping, but it's part of the RFD discussion.

For JSON formatting, I would like us to modify the JSON formatter such that trace errors are marshaled without all the escaping quotes, newlines, tabs, etc. I tested a patch that does that, it's actually a pretty simple change. Here's what it could look like in server logs with JSON formatting:

```
{
  "caller": "cache/cache.go:805",
  "component": "auth:1:cache",
  "error": {
    "message": "invalid origin value `bad-config`, must be one of [defaults config-file dynamic cloud kubernetes]",
    "user_messages": [
      "database `pg`",
      "failed to fetch resource: `db`"
    ]
  },
  "level": "warning",
  "message": "Cache `auth` first init failed, continuing re-init attempts in background.",
  "timestamp": "2023-01-31T14:26:52-08:00"
}
```

Aside from that though, why wrap the name in the caller?

Comment (greedy52) [reply api/types/database.go:492]: > why wrap the name in the caller?

It's easier to wrap there as it's a single place you have to add the name, and you can add some context too e.g. `trace.Wrap(err, "Failed to validate database resource %q",  database.Name)`. Just personal preference.

>```
>"user_messages": [
>      "database `pg`",
>      "failed to fetch resource: `db`"
>]
>```
I know the RFD is still under review, but the `database 'pg'` feels more like a field than a message. Might be awesome if we can do something like this in this case `trace.Wrap(err).WithField("db_name", db.GetName())` and the field logger can add the extra fields from the trace.

Again, I started this comment as I was curious about the new line. I don't have a strong opinion on the wrapping.

@smallinsky what do you think?

Comment (GavinFrazar) [reply api/types/database.go:492]: > It's easier to wrap there as it's a single place you have to add the name, and you can add some context too e.g. trace.Wrap(err, "Failed to validate database resource %q",  database.Name). Just personal preference.

It also looks more like a user message than a field, so this is probably the better way to go.

> we can do something like this in this case trace.Wrap(err).WithField("db_name", db.GetName()) and the field logger can add the extra fields from the trace.

There's a few places we already use `trace.WithField` and those fields are printed in debug mode with the text log formatter already as well. I'm in favor of adding those fields in the json formatter too, but i'm not sure exactly how we should handle that for user-facing logs in the CLI

Comment (smallinsky) [reply api/types/database.go:492]: A don't have strong opinion either but personally i would incline to the `trace.WithField` approach where the json and text formatter print output looks readable. We can do something like: `"log.WithFields(trace.Wrap(err).GetFields()).WithError(err).Warn("...")"`

**Review by gabrielcorado — COMMENTED** [review]
Given that we're validating the user options, we shouldn't have a scenario where user's input generate a bad `teleport.yaml`, right?

With that being said, I think it might be better to offer a secondary command to validate their configuration without needing to start Teleport, since the usual workflow might include users changing their configuration files. For example, having a `teleport db configure check teleport.yaml` command.

Still, this validation seems important on tests to ensure the config generator is up-to-date.

**Comment by GavinFrazar** [comment]
> Given that we're validating the user options, we shouldn't have a scenario where user's input generate a bad `teleport.yaml`, right?

yes

> With that being said, I think it might be better to offer a secondary command to validate their configuration without needing to start Teleport, since the usual workflow might include users changing their configuration files. For example, having a `teleport db configure check teleport.yaml` command.
> 
> Still, this validation seems important on tests to ensure the config generator is up-to-date.

I like the idea to add `teleport db configure check teleport.yaml`, but I don't think we should ever generate invalid config -- that should fail when user does `teleport db configure create`. If they later edit config and want to check it, then the command you suggest would be useful.

File: api/types/database.go
Comment (GavinFrazar) [reply api/types/database.go:492]: I think this thread is a problem with the way we format user messages from `trace.Wrap`, but that should be addressed as a separate issue specifically for how we format errors and logs.

If I use
```
 if err := db.CheckAndSetDefaults(); err != nil { 
 	return trace.Wrap(err, "failed to validate database %v resource", db.Name) 
```

Then the database name will be repeated in cases where we have errors that already embed the database name like `trace.BadParameter("database %q protocol is empty", d.GetName())`, and every log will include the newline/tabs with the way we format logs/errors right now.
Additionally, I'd have to find all callers of this function and also wrap the database name there to consistently provide the db name in the error message.

How about for now I'll just remove changes that added the wrapped user message to avoid inserting newlines in json error logs. I'd rather not block the PR on this issue

File: api/types/database.go
Comment (smallinsky) [inline api/types/database.go:527]: Why the error is wrapped with `trace.Wrap(err)` directly before `logrus.WithError(err)` call ?

**Review by smallinsky — COMMENTED** [review]
I have one question regarding trace.Wrap call but apart from that so LGTM when the comment will be resolved.

File: api/types/database.go
Comment (GavinFrazar) [reply api/types/database.go:527]: I thought we should wrap before logging the error, but I guess it doesn't matter since that func will have already wrapped all errors.

**Review by greedy52 — APPROVED** [review]
LGTM. i don't think the extra wrapping is needed before `logrus.WithError` either.

There could be some minor conflict with my other change #21480

File: api/types/database.go
Comment (GavinFrazar) [reply api/types/database.go]: removed wrapping

**Review by smallinsky — APPROVED** [review]

**Review by gabrielcorado — APPROVED** [review]

**Pull request merged**
